crates/fs/private/namefilter/bloomfilter.rs

```diff
@@ -0,0 +1,172 @@
+use std::ops::Index;
+
+use bitvec::prelude::BitArray;
+use xxhash_rust::xxh3;
+
+//------------------------------------------------------------------------------
+// Type Definitions
+//------------------------------------------------------------------------------
+
+/// The bloom filter is a probabilistic data structure that can be used to store a set of hashes.
+///
+/// `N` is the size of the bloom filter in bytes.
+///
+/// `K` is the number of bits to be set with each add operation.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct BloomFilter<const N: usize, const K: usize> {
+    pub(super) bits: BitArray<[u8; N]>,
+}
+
+/// An iterator that generates indices into some bloom filter based on deterministic hashing of specified item.
+///
+/// `N` is the number of bytes in the bloom filter.
+/// This is used to restrict generated value within bloomfilter index space bounds.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct HashIndexIterator<'a, T: AsRef<[u8]>, const N: usize> {
+    item: &'a T,
+    index: u64,
+}
+
+//------------------------------------------------------------------------------
+// Implementations
+//------------------------------------------------------------------------------
+
+impl<'a, T: AsRef<[u8]>, const N: usize> HashIndexIterator<'a, T, N> {
+    /// Creates a new iterator.
+    pub(super) fn new(item: &'a T) -> Self {
+        Self { item, index: 0 }
+    }
+
+    /// Returns the size of the bloom filter in bits.
+    pub const fn bit_size() -> usize {
+        N * 8
+    }
+}
+
+impl<T: AsRef<[u8]>, const N: usize> Iterator for HashIndexIterator<'_, T, N> {
+    type Item = usize;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let hash = xxh3::xxh3_64_with_seed(self.item.as_ref(), self.index) as usize;
+        let value = hash % Self::bit_size();
+        self.index += 1;
+        Some(value)
+    }
+}
+
+impl<const N: usize, const K: usize> BloomFilter<N, K> {
+    /// Creates a new bloom filter with all bits unset.
+    pub fn new() -> Self {
+        Self {
+            bits: Default::default(),
+        }
+    }
+
+    /// Adds an item to the bloom filter.
+    pub fn add<T>(&mut self, item: &T)
+    where
+        T: AsRef<[u8]>,
+    {
+        for i in self.hash_indices(item) {
+            self.bits.set(i, true);
+        }
+    }
+
+    /// Returns the number of hash iterations the bloom filter uses to set bits.
+    pub const fn bit_index_count(&self) -> usize {
+        K
+    }
+
+    /// Checks if the item is in the bloom filter.
+    pub fn contains<T>(&self, item: &T) -> bool
+    where
+        T: AsRef<[u8]>,
+    {
+        for i in self.hash_indices(item) {
+            if !self.bits[i] {
+                return false;
+            }
+        }
+
+        true
```

Comment:
I'm kind of codegolfing here, but how about this?
```suggestion
        self.hash_indices(item).all(|i| self.bits[i])
```